### DISCUSSION TITLE
Better synchronization function handling

### DISCUSSION THREAD

**Discussion by wu-clan** [fastapi-practices/fastapi_best_architecture#85] [Ideas]
Need to fix improper use of global synchronous or asynchronous functions

Introduce https://github.com/tiangolo/asyncer or a better tool or custom wrapper

**Comment by wu-clan** [comment]
Hi, @downdawn

Any thoughts on this idea, do you have any?

**Reply by downdawn** [reply]
I agree with your idea, and I believe it will lead to positive results.

**Comment by wu-clan** [comment]
The pursuit of simple and practical things:

1. https://github.com/pomponchik/awaits
2. https://github.com/Predeactor/ASyncTranslator
3. https://github.com/tiangolo/asyncer

My attitude, priority from top to bottom.

**Reply by wu-clan** [reply]
When using typing, [awaits](https://github.com/pomponchik/awaits) doesn't work very well in the IDE, e.g.: PyCharm

**Reply by wu-clan** [reply]
I used 2 and submitted a PR https://github.com/fastapi-practices/fastapi_best_architecture/pull/96, which looks like the original django implementation, but is asgi standard and seems to have no impact.

For approaches without large IO blocking, these transformations can seem senseless, but it's a good direction

**Reply by downdawn** [reply]
Why consider using ASyncTranslator instead of asyncer? The latter is more authoritative, the former looks a bit new?

**Reply by wu-clan** [reply]
asyncer is something more authoritative, but it makes things complicated

1. no parameter smart hints in PyCharm, which depends on the IDE does not support, and has not been fixed 
2. syntax is not concise enough

**Comment by wu-clan** [comment]
I decided to use https://github.com/django/asgiref

The source code of ASyncTranslator is from asgiref, just code separation without any highlights, but asgiref is the official implementation and application of django

**Reply by downdawn** [reply]
looks good!
